File: backend/app/schemas/desktop.py
"""Schemas for the desktop (display-manager) toggle feature."""
from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class DesktopState(str, Enum):
    RUNNING = "running"
    STOPPED = "stopped"
    UNKNOWN = "unknown"


class DesktopStatus(BaseModel):
    """Current state of the desktop display manager."""

    state: DesktopState
    display_manager: str = Field(description="Name of the display-manager unit, e.g. 'sddm'")
    detail: Optional[str] = Field(default=None, description="Optional human-readable detail")
    session_locked: Optional[bool] = Field(
        default=None,
        description=(
            "Whether the graphical session is locked; None when that cannot be "
            "determined. Filled in by the status route, not by the desktop "
            "backend - the status-bar collector polls the service every 10s and "
            "must not spawn a loginctl for a bit only the power menu reads."
        ),
    )

File: client/src/api/desktop.ts
/**
 * API client for Desktop (KDE/SDDM) toggle.
 *
 * Controls the display manager service so the GPU can enter a low-power
 * state while the NAS stays accessible. The desktop session is re-started
 * automatically when re-enabled.
 */

import { apiClient } from '../lib/api';

// ============================================================================
// Types
// ============================================================================

export type DesktopState = 'running' | 'stopped' | 'unknown';

export interface DesktopStatus {
  state: DesktopState;
  display_manager: string;
  detail: string | null;
  /**
   * Whether the graphical session is locked. `null` means the server could not
   * tell (no session, no loginctl) — never treat that as "unlocked".
   */
  session_locked?: boolean | null;
}

export interface DesktopActionResult {
  success: boolean;
  message: string;
  session_unlocked?: boolean;
  unlock_message?: string;
}

// ============================================================================
// API functions
// ============================================================================

export async function getDesktopStatus(): Promise<DesktopStatus> {
  const { data } = await apiClient.get<DesktopStatus>('/api/system/sleep/desktop/status');
  return data;
}

export async function disableDesktop(): Promise<DesktopActionResult> {
  const { data } = await apiClient.post<DesktopActionResult>('/api/system/sleep/desktop/disable');
  return data;
}

export async function enableDesktop(): Promise<DesktopActionResult> {
  const { data } = await apiClient.post<DesktopActionResult>('/api/system/sleep/desktop/enable');
  return data;
}

/**
 * Unlock the KDE session without touching the displays.
 *
 * `enableDesktop()` unlocks as a side effect of turning the screens on; this is
 * the same action for when they are already on. A refusal (wrong network, no
 * permission) comes back as `success: false`, not as an HTTP error.
 */
export async function unlockSession(): Promise<DesktopActionResult> {
  const { data } = await apiClient.post<DesktopActionResult>('/api/system/sleep/desktop/unlock');
  return data;
}
